tests/Unit/MicroPlateTest.php

```diff
@@ -25,11 +28,126 @@ public function testCanAddAndRetriedWellBasedOnCoordinateSystem(): void
+        self::assertSame('A1', $sortedByColumn->keys()->first());
+        self::assertSame('B1', $sortedByColumn->keys()->skip(1)->first());
+        self::assertSame('C1', $sortedByColumn->keys()->skip(2)->first());
+        self::assertSame('D1', $sortedByColumn->keys()->skip(3)->first());
+        self::assertSame('H12', $sortedByColumn->keys()->skip(95)->last());
+        self::assertSame('G12', $sortedByColumn->keys()->skip(94)->first());
+    }
+
+    public function testFreeWells(): void
+    {
+        $microPlate = $this->perparePlate();
+
+        self::assertTrue(
+            $microPlate->wells->some(fn ($value, $key) => null !== $value)
+        );
+        $microPlate->freeWells()->map(fn ($value, $key) => self::assertNull($value));
+    }
+
+    /**
+     * @phpstan-return MicroPlate<mixed, CoordinateSystem96Well>
+     */
+    private function perparePlate(): MicroPlate
+    {
+        $coordinateSystem = new CoordinateSystem96Well();
+
+        $microPlate = new MicroPlate($coordinateSystem);
+
+        $dataProvider96Well = CoordinateTest::dataProvider96Well();
+        shuffle($dataProvider96Well);
+        foreach ($dataProvider96Well as $wellData) {
+            $randomNumber = rand(1, 100);
+            $microPlateCoordinate = new Coordinate($wellData['row'], $wellData['column'], $coordinateSystem);
+            $microPlate->addWell($microPlateCoordinate, $randomNumber > 50 ? $randomNumber : null);
+        }
+
+        return $microPlate;
+    }
+
+    public function testNextFreeWellAddingAndGetting(): void
+    {
+        $coordinateSystem = new CoordinateSystem96Well();
+        $microPlate = new MicroPlate($coordinateSystem);
+
+        $microPlateCoordinate1 = new Coordinate('A', 1, $coordinateSystem);
+        self::assertEquals($microPlateCoordinate1, $microPlate->nextFreeWellCoordinate(FlowDirection::COLUMN()));
+        $microPlate->addToNextFreeWell('foo', FlowDirection::COLUMN());
+
+        $microPlateCoordinate2 = new Coordinate('B', 1, $coordinateSystem);
+        self::assertEquals($microPlateCoordinate2, $microPlate->nextFreeWellCoordinate(FlowDirection::COLUMN()));
+        $microPlate->addToNextFreeWell('bar', FlowDirection::COLUMN());
+
+        $microPlateCoordinate3 = new Coordinate('C', 1, $coordinateSystem);
+        self::assertEquals($microPlateCoordinate3, $microPlate->nextFreeWellCoordinate(FlowDirection::COLUMN()));
+
+        $microPlateCoordinate4 = new Coordinate('A', 2, $coordinateSystem);
+        self::assertEquals($microPlateCoordinate4, $microPlate->nextFreeWellCoordinate(FlowDirection::ROW()));
+        $microPlate->addToNextFreeWell('foobar', FlowDirection::ROW());
+
+        $microPlateCoordinate5 = new Coordinate('A', 3, $coordinateSystem);
+        self::assertEquals($microPlateCoordinate5, $microPlate->nextFreeWellCoordinate(FlowDirection::ROW()));
+        $microPlate->addToNextFreeWell('barfoo', FlowDirection::ROW());
+
+        self::assertSame([
+            'A1' => 'foo',
```

Comment:
Where does `'foo'` come from? Use variables to deduplicate and show relation between input and expected result